apps/ex_nvr/lib/ex_nvr/remote_storages/snapshot_uploader.ex

```diff
@@ -0,0 +1,128 @@
+defmodule ExNVR.RemoteStorages.SnapshotUploader do
+  use GenServer, restart: :transient
+
+  require Logger
+
+  alias ExNVR.RemoteStorage
+  alias ExNVR.{Devices, RemoteStorages}
+  alias ExNVR.RemoteStorages.Store
+
+  def start_link(options) do
+    GenServer.start_link(__MODULE__, options)
+  end
+
+  @impl true
+  def init(options) do
+    Logger.metadata(device_id: options[:device].id)
+    Logger.info("Start snapshot uploader")
+    send(self(), :init_config)
+    {:ok, %{device: options[:device], schedule: %{}, opts: []}}
+  end
+
+  @impl true
+  def handle_info(:init_config, %{device: device} = state) do
+    snapshot_config = device.snapshot_config
+    stream_config = device.stream_config
+
+    with false <- is_nil(stream_config.snapshot_uri),
+         true <- snapshot_config.enabled,
+         %RemoteStorage{} = remote_storage <-
+           RemoteStorages.get_by(name: snapshot_config.remote_storage) do
+      schedule = parse_schedule(device)
+      opts = build_opts(remote_storage)
+      send(self(), :upload_snapshot)
+      {:noreply, %{state | schedule: schedule, opts: opts}}
+    else
+      _ ->
+        Logger.info("Stop snapshot uploader")
+        {:stop, :normal, state}
+    end
+  end
+
+  @impl true
+  def handle_info(:upload_snapshot, %{device: device, schedule: schedule, opts: opts} = state) do
+    snapshot_config = device.snapshot_config
+    utc_now = DateTime.utc_now()
+
+    with true <- scheduled?(device, schedule),
+         {:ok, snapshot} <- Devices.fetch_snapshot(device) do
```

Comment:
Add a timeout to this request less than the interval and a max of `15 seconds` if the interval is big.